FreeBSD/usr/src/share/man/man4/icmp.4
.\" Copyright (c) 1986, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)icmp.4	8.1 (Berkeley) 6/5/93
.\" $FreeBSD: releng/11.1/share/man/man4/icmp.4 280688 2015-03-26 12:52:21Z ru $
.\"
.Dd March 26, 2015
.Dt ICMP 4
.Os
.Sh NAME
.Nm icmp
.Nd Internet Control Message Protocol
.Sh SYNOPSIS
.In sys/types.h
.In sys/socket.h
.In netinet/in.h
.Ft int
.Fn socket AF_INET SOCK_RAW proto
.Sh DESCRIPTION
.Tn ICMP
is the error and control message protocol used
by
.Tn IP
and the Internet protocol family.
It may be accessed
through a
.Dq raw socket
for network monitoring
and diagnostic functions.
The
.Fa proto
parameter to the socket call to create an
.Tn ICMP
socket
is obtained from
.Xr getprotobyname 3 .
.Tn ICMP
sockets are connectionless,
and are normally used with the
.Xr sendto 2
and
.Xr recvfrom 2
calls, though the
.Xr connect 2
call may also be used to fix the destination for future
packets (in which case the
.Xr read 2
or
.Xr recv 2
and
.Xr write 2
or
.Xr send 2
system calls may be used).
.Pp
Outgoing packets automatically have an
.Tn IP
header prepended to
them (based on the destination address).
Incoming packets are received with the
.Tn IP
header and options intact.
.Ss Types
ICMP messages are classified according to the type and code fields
present in the ICMP header.
The abbreviations for the types and codes may be used in rules in
.Xr pf.conf 5 .
The following types are defined:
.Bl -column x xxxxxxxxxxxx -offset indent
.It Sy Num Ta Sy Abbrev. Ta Sy Description
.It 0 Ta echorep Ta "Echo reply"
.It 3 Ta unreach Ta "Destination unreachable"
.It 4 Ta squench Ta "Packet loss, slow down"
.It 5 Ta redir Ta "Shorter route exists"
.It 6 Ta althost Ta "Alternate host address"
.It 8 Ta echoreq Ta "Echo request"
.It 9 Ta routeradv Ta "Router advertisement"
.It 10 Ta routersol Ta "Router solicitation"
.It 11 Ta timex Ta "Time exceeded"
.It 12 Ta paramprob Ta "Invalid IP header"
.It 13 Ta timereq Ta "Timestamp request"
.It 14 Ta timerep Ta "Timestamp reply"
.It 15 Ta inforeq Ta "Information request"
.It 16 Ta inforep Ta "Information reply"
.It 17 Ta maskreq Ta "Address mask request"
.It 18 Ta maskrep Ta "Address mask reply"
.It 30 Ta trace Ta Traceroute
.It 31 Ta dataconv Ta "Data conversion problem"
.It 32 Ta mobredir Ta "Mobile host redirection"
.It 33 Ta ipv6-where Ta "IPv6 where-are-you"
.It 34 Ta ipv6-here Ta "IPv6 i-am-here"
.It 35 Ta mobregreq Ta "Mobile registration request"
.It 36 Ta mobregrep Ta "Mobile registration reply"
.It 39 Ta skip Ta SKIP
.It 40 Ta photuris Ta Photuris
.El
.Pp
The following codes are defined:
.Bl -column x xxxxxxxxxxxx xxxxxxxx -offset indent
.It Sy Num Ta Sy Abbrev. Ta Sy Type Ta Sy Description
.It 0 Ta net-unr Ta unreach Ta "Network unreachable"
.It 1 Ta host-unr Ta unreach Ta "Host unreachable"
.It 2 Ta proto-unr Ta unreach Ta "Protocol unreachable"
.It 3 Ta port-unr Ta unreach Ta "Port unreachable"
.It 4 Ta needfrag Ta unreach Ta "Fragmentation needed but DF bit set"
.It 5 Ta srcfail Ta unreach Ta "Source routing failed"
.It 6 Ta net-unk Ta unreach Ta "Network unknown"
.It 7 Ta host-unk Ta unreach Ta "Host unknown"
.It 8 Ta isolate Ta unreach Ta "Host isolated"
.It 9 Ta net-prohib Ta unreach Ta "Network administratively prohibited"
.It 10 Ta host-prohib Ta unreach Ta "Host administratively prohibited"
.It 11 Ta net-tos Ta unreach Ta "Invalid TOS for network"
.It 12 Ta host-tos Ta unreach Ta "Invalid TOS for host"
.It 13 Ta filter-prohib Ta unreach Ta "Prohibited access"
.It 14 Ta host-preced Ta unreach Ta "Precedence violation"
.It 15 Ta cutoff-preced Ta unreach Ta "Precedence cutoff"
.It 0 Ta redir-net Ta redir Ta "Shorter route for network"
.It 1 Ta redir-host Ta redir Ta "Shorter route for host"
.It 2 Ta redir-tos-net Ta redir Ta "Shorter route for TOS and network"
.It 3 Ta redir-tos-host Ta redir Ta "Shorter route for TOS and host"
.It 0 Ta normal-adv Ta routeradv Ta "Normal advertisement"
.It 16 Ta common-adv Ta routeradv Ta "Selective advertisement"
.It 0 Ta transit Ta timex Ta "Time exceeded in transit"
.It 1 Ta reassemb Ta timex Ta "Time exceeded in reassembly"
.It 0 Ta badhead Ta paramprob Ta "Invalid option pointer"
.It 1 Ta optmiss Ta paramprob Ta "Missing option"
.It 2 Ta badlen Ta paramprob Ta "Invalid length"
.It 1 Ta unknown-ind Ta photuris Ta "Unknown security index"
.It 2 Ta auth-fail Ta photuris Ta "Authentication failed"
.It 3 Ta decrypt-fail Ta photuris Ta "Decryption failed"
.El
.Ss MIB Variables
The
.Tn ICMP
protocol implements a number of variables in the
.Va net.inet.icmp
branch of the
.Xr sysctl 3
MIB.
.Bl -tag -width ".Va icmplim_output"
.It Va maskrepl
.Pq Vt boolean
Enable/disable replies to ICMP Address Mask Request packets.
Defaults to false.
.It Va maskfake
.Pq Vt "unsigned integer"
When
.Va maskrepl
is set and this value is non-zero,
it will be used instead of the real address mask when
the system replies to an ICMP Address Mask Request packet.
Defaults to 0.
.It Va icmplim
.Pq Vt integer
Bandwidth limit for ICMP replies in packets/second.
If set to zero, no limiting will occur.
Defaults to 200.
.It Va icmplim_output
.Pq Vt boolean
Enable/disable logging of ICMP replies bandwidth limiting.
Defaults to true.
.It Va drop_redirect
.Pq Vt boolean
Enable/disable dropping of ICMP Redirect packets.
Defaults to false.
.It Va log_redirect
.Pq Vt boolean
Enable/disable logging of ICMP Redirect packets.
Defaults to false.
.It Va bmcastecho
.Pq Vt boolean
Enable/disable ICMP replies received via broadcast or multicast.
Defaults to false.
.It Va reply_src
.Pq Vt str
An interface name used for the ICMP reply source in response to packets
which are not directly addressed to us.
By default continue with normal source selection.
.It Va reply_from_interface
.Pq Vt boolean
Use the IP address of the interface the packet came through in for
responses to packets which are not directly addressed to us.
If enabled, this rule is processed before all others.
By default, continue with normal source selection.
Enabling this option is particularly useful on routers because it
makes external traceroutes show the actual path a packet has taken
instead of the possibly different return path.
.It Va quotelen
.Pq Vt integer
Number of bytes from original packet to quote in ICMP reply.
This number is internally enforced to be at least 8 bytes (per RFC792)
and at most the maximal space left in the ICMP reply mbuf.
.It Va tstamprepl
.Pq Vt boolean
Enable/disable replies to ICMP Timestamp packets.
Defaults to true.
.El
.Sh ERRORS
A socket operation may fail with one of the following errors returned:
.Bl -tag -width Er
.It Bq Er EISCONN
when trying to establish a connection on a socket which
already has one, or when trying to send a datagram with the destination
address specified and the socket is already connected;
.It Bq Er ENOTCONN
when trying to send a datagram, but
no destination address is specified, and the socket has not been
connected;
.It Bq Er ENOBUFS
when the system runs out of memory for
an internal data structure;
.It Bq Er EADDRNOTAVAIL
when an attempt is made to create a
socket with a network address for which no network interface
exists.
.El
.Sh SEE ALSO
.Xr recv 2 ,
.Xr send 2 ,
.Xr inet 4 ,
.Xr intro 4 ,
.Xr ip 4 ,
.Xr pf.conf 5
.Sh HISTORY
The
.Nm
protocol appeared in
.Bx 4.3 .


FreeBSD/usr/src/usr.bin/systat/icmp.c
/*-
 * Copyright (c) 1980, 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>

__FBSDID("$FreeBSD: releng/11.1/usr.bin/systat/icmp.c 240605 2012-09-17 13:36:47Z melifaro $");

#ifdef lint
static char sccsid[] = "@(#)mbufs.c	8.1 (Berkeley) 6/6/93";
#endif

/* From:
	"Id: mbufs.c,v 1.5 1997/02/24 20:59:03 wollman Exp"
*/

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sysctl.h>

#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <netinet/icmp_var.h>

#include <stdlib.h>
#include <string.h>
#include <paths.h>
#include "systat.h"
#include "extern.h"
#include "mode.h"

static struct icmpstat icmpstat, initstat, oldstat;

/*-
--0         1         2         3         4         5         6         7
--0123456789012345678901234567890123456789012345678901234567890123456789012345
00          ICMP Input                         ICMP Output
01999999999 total messages           999999999 total messages
02999999999 with bad code            999999999 errors generated
03999999999 with bad length          999999999 suppressed - original too short
04999999999 with bad checksum        999999999 suppressed - original was ICMP
05999999999 with insufficient data   999999999 responses sent
06                                   999999999 suppressed - multicast echo
07                                   999999999 suppressed - multicast tstamp
08
09          Input Histogram                    Output Histogram
10999999999 echo response            999999999 echo response
11999999999 echo request             999999999 echo request
12999999999 destination unreachable  999999999 destination unreachable
13999999999 redirect                 999999999 redirect
14999999999 time-to-live exceeded    999999999 time-to-line exceeded
15999999999 parameter problem        999999999 parameter problem
16999999999 router advertisement     999999999 router solicitation
17
18
--0123456789012345678901234567890123456789012345678901234567890123456789012345
--0         1         2         3         4         5         6         7
*/

WINDOW *
openicmp(void)
{
	return (subwin(stdscr, LINES-3-1, 0, MAINWIN_ROW, 0));
}

void
closeicmp(WINDOW *w)
{
	if (w == NULL)
		return;
	wclear(w);
	wrefresh(w);
	delwin(w);
}

void
labelicmp(void)
{
	wmove(wnd, 0, 0); wclrtoeol(wnd);
#define L(row, str) mvwprintw(wnd, row, 10, str)
#define R(row, str) mvwprintw(wnd, row, 45, str);
	L(0, "ICMP Input");		R(0, "ICMP Output");
	L(1, "total messages");		R(1, "total messages");
	L(2, "with bad code");		R(2, "errors generated");
	L(3, "with bad length");	R(3, "suppressed - original too short");
	L(4, "with bad checksum");	R(4, "suppressed - original was ICMP");
	L(5, "with insufficient data");	R(5, "responses sent");
					R(6, "suppressed - multicast echo");
					R(7, "suppressed - multicast tstamp");
	L(9, "Input Histogram");	R(9, "Output Histogram");
#define B(row, str) L(row, str); R(row, str)
	B(10, "echo response");
	B(11, "echo request");
	B(12, "destination unreachable");
	B(13, "redirect");
	B(14, "time-to-live exceeded");
	B(15, "parameter problem");
	L(16, "router advertisement");	R(16, "router solicitation");
#undef L
#undef R
#undef B
}

static void
domode(struct icmpstat *ret)
{
	const struct icmpstat *sub;
	int i, divisor = 1;

	switch(currentmode) {
	case display_RATE:
		sub = &oldstat;
		divisor = (delay > 1000000) ? delay / 1000000 : 1;
		break;
	case display_DELTA:
		sub = &oldstat;
		break;
	case display_SINCE:
		sub = &initstat;
		break;
	default:
		*ret = icmpstat;
		return;
	}
#define DO(stat) ret->stat = (icmpstat.stat - sub->stat) / divisor
	DO(icps_error);
	DO(icps_oldshort);
	DO(icps_oldicmp);
	for (i = 0; i <= ICMP_MAXTYPE; i++) {
		DO(icps_outhist[i]);
	}
	DO(icps_badcode);
	DO(icps_tooshort);
	DO(icps_checksum);
	DO(icps_badlen);
	DO(icps_reflect);
	for (i = 0; i <= ICMP_MAXTYPE; i++) {
		DO(icps_inhist[i]);
	}
	DO(icps_bmcastecho);
	DO(icps_bmcasttstamp);
#undef DO
}

void
showicmp(void)
{
	struct icmpstat stats;
	u_long totalin, totalout;
	int i;

	memset(&stats, 0, sizeof stats);
	domode(&stats);
	for (i = totalin = totalout = 0; i <= ICMP_MAXTYPE; i++) {
		totalin += stats.icps_inhist[i];
		totalout += stats.icps_outhist[i];
	}
	totalin += stats.icps_badcode + stats.icps_badlen +
		stats.icps_checksum + stats.icps_tooshort;
	mvwprintw(wnd, 1, 0, "%9lu", totalin);
	mvwprintw(wnd, 1, 35, "%9lu", totalout);

#define DO(stat, row, col) \
	mvwprintw(wnd, row, col, "%9lu", stats.stat)

	DO(icps_badcode, 2, 0);
	DO(icps_badlen, 3, 0);
	DO(icps_checksum, 4, 0);
	DO(icps_tooshort, 5, 0);
	DO(icps_error, 2, 35);
	DO(icps_oldshort, 3, 35);
	DO(icps_oldicmp, 4, 35);
	DO(icps_reflect, 5, 35);
	DO(icps_bmcastecho, 6, 35);
	DO(icps_bmcasttstamp, 7, 35);
#define DO2(type, row) DO(icps_inhist[type], row, 0); DO(icps_outhist[type], \
							 row, 35)
	DO2(ICMP_ECHOREPLY, 10);
	DO2(ICMP_ECHO, 11);
	DO2(ICMP_UNREACH, 12);
	DO2(ICMP_REDIRECT, 13);
	DO2(ICMP_TIMXCEED, 14);
	DO2(ICMP_PARAMPROB, 15);
	DO(icps_inhist[ICMP_ROUTERADVERT], 16, 0);
	DO(icps_outhist[ICMP_ROUTERSOLICIT], 16, 35);
#undef DO
#undef DO2
}

int
initicmp(void)
{
	size_t len;
	int name[4];

	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_ICMP;
	name[3] = ICMPCTL_STATS;

	len = 0;
	if (sysctl(name, 4, 0, &len, 0, 0) < 0) {
		error("sysctl getting icmpstat size failed");
		return 0;
	}
	if (len > sizeof icmpstat) {
		error("icmpstat structure has grown--recompile systat!");
		return 0;
	}
	if (sysctl(name, 4, &initstat, &len, 0, 0) < 0) {
		error("sysctl getting icmpstat size failed");
		return 0;
	}
	oldstat = initstat;
	return 1;
}

void
reseticmp(void)
{
	size_t len;
	int name[4];

	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_ICMP;
	name[3] = ICMPCTL_STATS;

	len = sizeof initstat;
	if (sysctl(name, 4, &initstat, &len, 0, 0) < 0) {
		error("sysctl getting icmpstat size failed");
	}
	oldstat = initstat;
}

void
fetchicmp(void)
{
	int name[4];
	size_t len;

	oldstat = icmpstat;
	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_ICMP;
	name[3] = ICMPCTL_STATS;
	len = sizeof icmpstat;

	if (sysctl(name, 4, &icmpstat, &len, 0, 0) < 0)
		return;
}
